src/flb_trace_chunk.c

```diff
@@ -0,0 +1,654 @@
+static inline int flb_trace_chunk_to_be_destroyed(struct flb_trace_chunk_context *ctxt)
+{
+    int ret = FLB_FALSE;
+
+    ret = (ctxt->to_destroy == 1 ? FLB_TRUE : FLB_FALSE);
+    return ret;
+}
+
+static inline int flb_trace_chunk_has_chunks(struct flb_trace_chunk_context *ctxt)
+{
+    int ret = FLB_FALSE;
+
+    ret = ((ctxt->chunks > 0) ? FLB_TRUE : FLB_FALSE);
+    return ret;
+}
+
+static inline void flb_trace_chunk_add(struct flb_trace_chunk_context *ctxt)
+{
+    ctxt->chunks++;
+}
+
+static inline void flb_trace_chunk_sub(struct flb_trace_chunk_context *ctxt)
+{
+    ctxt->chunks--;
+}
+
+static inline void flb_trace_chunk_set_destroy(struct flb_trace_chunk_context *ctxt)
+{
+    ctxt->to_destroy = 1;
+}
+
+static struct flb_output_instance *find_calyptia_output_instance(struct flb_config *config)
+{
+    struct mk_list *head;
+    struct flb_output_instance *output;
+
+    mk_list_foreach(head, &config->outputs) {
+        output = mk_list_entry(head, struct flb_output_instance, _head);
+        if (strcmp(output->p->name, "calyptia") == 0) {
+            return output;
+        }
+    }
+    return NULL;
+}
+
+static void log_cb(void *ctx, int level, const char *file, int line,
+                   const char *str)
+{
+    if (level == CIO_LOG_ERROR) {
+        flb_error("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_WARN) {
+        flb_warn("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_INFO) {
+        flb_info("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_DEBUG) {
+        flb_debug("[trace] %s", str);
+    }
+}
+
+static void trace_chunk_context_destroy(struct flb_trace_chunk_context *ctxt)
+{
+    if (flb_trace_chunk_has_chunks(ctxt) == FLB_TRUE) {
+        flb_trace_chunk_set_destroy(ctxt);
+        flb_input_pause_all(ctxt->flb->config);
+        return;
+    }
+    
+    // pause all inputs, then destroy the input storage.
+    flb_input_pause_all(ctxt->flb->config);
+    // waiting for all tasks to end is key to safely stopping and destroying
+    // the fluent-bit pipeline.
+    while (flb_task_running_count(ctxt->flb->config) > 0) {
+        sleep(1);
+    }
+    flb_sds_destroy(ctxt->trace_prefix);
+    flb_stop(ctxt->flb);
+    flb_destroy(ctxt->flb);
+    flb_free(ctxt);
+}
+
+void flb_trace_chunk_context_destroy(void *input)
+{
+    struct flb_input_instance *in = (struct flb_input_instance *)input;
+    pthread_mutex_lock(&in->trace_lock);
+    if (in->trace_ctxt != NULL) {
+        trace_chunk_context_destroy(in->trace_ctxt);
+        in->trace_ctxt = NULL;
+    }
+    pthread_mutex_unlock(&in->trace_lock);
+}
+
+struct flb_trace_chunk_context *flb_trace_chunk_context_new(void *trace_input, const char *output_name, const char *trace_prefix, void *data, struct mk_list *props)
+{
+    struct flb_input_instance *in = (struct flb_input_instance *)trace_input;
+    struct flb_config *config = in->config;
+    pthread_mutexattr_t attr = {0};
+    struct flb_input_instance *input;
```

Comment:
initialize all this variables as NULL (just in case some check assumes something)